File: ios/Classes/readers/configurations/PlatformContextPropertiesReader.swift
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

import Foundation
import SnowplowTracker

struct PlatformContextPropertiesReader: Decodable {
    let osType: String?
    let osVersion: String?
    let deviceVendor: String?
    let deviceModel: String?
    let carrier: String?
    let networkType: String?
    let networkTechnology: String?
    let appleIdfa: String?
    let appleIdfv: String?
    let availableStorage: String?;
    let totalStorage: String?
    let physicalMemory: String?
    let appAvailableMemory: Int?
    let batteryLevel: Int?
    let batteryState: String?
    let lowPowerMode: Bool?
    let isPortrait: Bool?
    let resolution: String?
    let scale: Double?
    let language: String?
}

extension PlatformContextPropertiesReader {
    func toPlatformContextRetriever() -> PlatformContextRetriever {
        let retriever = PlatformContextRetriever()

        if let osType = self.osType { retriever.osType = { osType } }
        if let osVersion = self.osVersion { retriever.osVersion = { osVersion } }
        if let deviceVendor = self.deviceVendor { retriever.deviceVendor = { deviceVendor } }
        if let deviceModel = self.deviceModel { retriever.deviceModel = { deviceModel } }
        if let carrier = self.carrier { retriever.carrier = { carrier } }
        if let networkType = self.networkType { retriever.networkType = { networkType } }
        if let networkTechnology = self.networkTechnology { retriever.networkTechnology = { networkTechnology } }
        if let appleIdfa = self.appleIdfa {
            if let uuid = UUID(uuidString: appleIdfa) {
                retriever.appleIdfa = { uuid }
            }
        }
        if let appleIdfv = self.appleIdfv { retriever.appleIdfv = { appleIdfv } }
        if let availableStorage = self.availableStorage { retriever.availableStorage = { Int64(availableStorage) } }
        if let totalStorage = self.totalStorage { retriever.totalStorage = { Int64(totalStorage) } }
        if let physicalMemory = self.physicalMemory { retriever.physicalMemory = { UInt64(physicalMemory) } }
        if let appAvailableMemory = self.appAvailableMemory { retriever.appAvailableMemory = { appAvailableMemory } }
        if let batteryLevel = self.batteryLevel { retriever.batteryLevel = { batteryLevel } }
        if let batteryState = self.batteryState { retriever.batteryState = { batteryState } }
        if let lowPowerMode = self.lowPowerMode { retriever.lowPowerMode = { lowPowerMode } }
        if let isPortrait = self.isPortrait { retriever.isPortrait = { isPortrait } }
        if let resolution = self.resolution { retriever.resolution = { resolution } }
        if let scale = self.scale { retriever.scale = { scale } }
        if let language = self.language { retriever.language = { language } }

        return retriever
    }
}

File: android/src/main/kotlin/com/snowplowanalytics/snowplow_tracker/readers/configurations/PlatformContextPropertiesReader.kt
// Copyright (c) 2022-present Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_tracker.readers.configurations

import android.content.Context
import com.snowplowanalytics.snowplow.configuration.TrackerConfiguration
import com.snowplowanalytics.snowplow.tracker.DevicePlatform
import com.snowplowanalytics.snowplow.tracker.PlatformContextRetriever
import com.snowplowanalytics.snowplow_tracker.TrackerVersion

class PlatformContextPropertiesReader(values: Map<String, Any>) {
    private val valuesDefault = values.withDefault { null }

    val osType: String? by valuesDefault
    val osVersion: String? by valuesDefault
    val deviceVendor: String? by valuesDefault
    val deviceModel: String? by valuesDefault
    val carrier: String? by valuesDefault
    val networkType: String? by valuesDefault
    val networkTechnology: String? by valuesDefault
    val androidIdfa: String? by valuesDefault
    val availableStorage: String? by valuesDefault
    val totalStorage: String? by valuesDefault
    val physicalMemory: String? by valuesDefault
    val systemAvailableMemory: String? by valuesDefault
    val batteryLevel: Int? by valuesDefault
    val batteryState: String? by valuesDefault
    val isPortrait: Boolean? by valuesDefault
    val resolution: String? by valuesDefault
    val scale: Double? by valuesDefault
    val language: String? by valuesDefault
    val appSetId: String? by valuesDefault
    val appSetIdScope: String? by valuesDefault

    fun toPlatformContextRetriever(): PlatformContextRetriever {
        val retriever = PlatformContextRetriever()

        osType?.let { retriever.osType = { it } }
        osVersion?.let { retriever.osVersion = { it } }
        deviceVendor?.let { retriever.deviceVendor = { it } }
        deviceModel?.let { retriever.deviceModel = { it } }
        carrier?.let { retriever.carrier = { it } }
        networkType?.let { retriever.networkType = { it } }
        networkTechnology?.let { retriever.networkTechnology = { it } }
        androidIdfa?.let { retriever.androidIdfa = { it } }
        availableStorage?.let { retriever.availableStorage = { it.toLong() } }
        totalStorage?.let { retriever.totalStorage = { it.toLong() } }
        physicalMemory?.let { retriever.physicalMemory = { it.toLong() } }
        systemAvailableMemory?.let { retriever.systemAvailableMemory = { it.toLong() } }
        batteryLevel?.let { retriever.batteryLevel = { it } }
        batteryState?.let { retriever.batteryState = { it } }
        isPortrait?.let { retriever.isPortrait = { it } }
        resolution?.let { retriever.resolution = { it } }
        scale?.let { retriever.scale = { it.toFloat() } }
        language?.let { retriever.language = { it } }
        appSetId?.let { retriever.appSetId = { it } }
        appSetIdScope?.let { retriever.appSetIdScope = { it } }

        return retriever
    }
}
